### ISSUE TITLE
Shell sessions running lazydocker over SSH become unusable after losing the SSH connection

### ISSUE THREAD

**Issue by zheng** [warpdotdev/Warp#4870]
### Dupe Check

- [X] I have [searched Warp bugs](https://github.com/warpdotdev/warp/issues?q=is%3Aissue+label%3ABugs) and there are no duplicates

### Describe the bug

<img width="863" alt="Screenshot 2024-05-03 at 2 33 50 PM" src="https://github.com/warpdotdev/Warp/assets/29991019/c02eea6e-38f8-4c1c-99da-5490f45a4841">
<img width="874" alt="Screenshot 2024-05-03 at 2 33 41 PM" src="https://github.com/warpdotdev/Warp/assets/29991019/58af93b7-f638-4bc4-929a-5e660244bc54">

Whenever the SSH session disconnects (e.g. when my laptop sleeps), I get these escape sequences that spam the session and I can never recover the session. I need to close the pane and then create a new one

### To reproduce

I haven't invested in reproducing this situation. If there's appetite to fix this, I'm happy to try and create a VM, or screen share to debug this

### Operating system

MacOS

### Operating system and version

Sonoma 14.3

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Does this block you from using Warp daily?

Yes, this issue prevents me from using Warp daily.

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

Yes, this I confirmed this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by alokedesai** [comment]
Miss you Zheng!

**Comment by zheng** [comment]
Miss you too!

By the way, the subshell bootstrapping works **incredibly well**. The team did an incredible job on that. I've been SSH-ing every session and I'm amazed at how well it works

**Comment by alokedesai** [comment]
We're actively working on a set of a improvements to improve SSH, including these SSH hangup issues. Should have an update in the next few weeks and we may reach out if we need any help reproducing (if that's ok with you!)

**Comment by zheng** [comment]
> We're actively working on a set of a improvements to improve SSH, including these SSH hangup issues. Should have an update in the next few weeks and we may reach out if we need any help reproducing (if that's ok with you!)

Yes, please reach out if I can be helpful for this or any other ssh issues!

**Comment by vorporeal** [comment]
I believe I have a fix for this!

@zheng As a quick test, can you reproduce this state, and then:

1. Press ctrl+u to clear anything in the input buffer
2. Run `vim`
3. Quit `vim`

and let me know if that gets you back into a good state?

I believe the issue here is that we're not forcibly exiting the alt screen when we get a notification from the shell that a command finished (and `ssh` isn't sending us any escape sequence to exit the alt screen either).  We end up stuck in the alt screen (fixed by running and quitting vim), and so control sequences that are intended to be consumed by alt screen programs (e.g.: telling a TUI program about mouse movements) are actually getting inserted into the shell's line editor.

Will have a fix out in next week's release. :)

**Comment by vorporeal** [comment]
Reopening this until it goes live (but the fix is merged).

**Comment by zheng** [comment]
> I believe I have a fix for this!
> 
> @zheng As a quick test, can you reproduce this state, and then:
> 
> 1. Press ctrl+u to clear anything in the input buffer
> 2. Run `vim`
> 3. Quit `vim`
> 
> and let me know if that gets you back into a good state?
> 
> I believe the issue here is that we're not forcibly exiting the alt screen when we get a notification from the shell that a command finished (and `ssh` isn't sending us any escape sequence to exit the alt screen either). We end up stuck in the alt screen (fixed by running and quitting vim), and so control sequences that are intended to be consumed by alt screen programs (e.g.: telling a TUI program about mouse movements) are actually getting inserted into the shell's line editor.
> 
> Will have a fix out in next week's release. :)

I've opened a SSH session to reproduce this state, should be able to try out this fix in the next few hours!

**Comment by alokedesai** [comment]
@zheng Thank you! The easiest way we reproduced was by temporarily turning off wifi (in case it helps you repro more reliably)

**Comment by zheng** [comment]
The fix works!
